core/src/main/scala/org/apache/spark/deploy/DeployMessage.scala

```diff
@@ -110,6 +122,16 @@ private[deploy] object DeployMessages {
 
   case class ReconnectWorker(masterUrl: String) extends DeployMessage
 
+  /**
+   * Ask the worker to release the indicated resources in ALLOCATED_RESOURCES_JSON_FILE
+   * @param pid process id of the target worker
+   * @param toRelease the resources expected to release
+   */
+  case class ReleaseResources(
```

Comment:
This is for a possible corner case that a worker lost with master but still alive. In this case, pid cleaning seems doesn't work, right ?

Reply:
You are saying a worker is still alive but something caused the worker to disconnect from Master.  

so there are 2 cases:
1) another worker exists and we can send this message to it to try to free those resources. But what if this Worker is still alive and whatever its running is still using those resources?  like ML algo using the GPU.  Do we really want the other worker to say their free if its potentially being used.
2) no other worker exists on that node so this message isn't sent anyway.

There are obviously many cases the worker could disconnect. If the process is still in a sane state then it will either reregister with master or eventually fail to contact master and release the resources at that point.    I think in both of those cases the Executors should have been killed or similarly failed due to too many heartbeat failures, correct?  At least the master was told they were lost.  So in those cases if we just wait then the resources should be cleaned up without us fixing anything.  The other case is worker isn't alive, this then would be covered by our PID checks, the last case is the worker and possibly executors are in some really bad state such that they are alive but aren't connected and heartbeat failures aren't causing them to exit.  In this case I would say its really had to say if its using the resources and I'm not sure we want to reclaim them automatically. It might be best to leave them still allocated until someone can clean them up.

Reply:
Make sense. I didn't think this part clearly before.

> Do we really want the other worker to say their free if its potentially being used.

But this is one thing I'd like to clarify that `ReleaseResources` only release those free resource of that lost worker and that's what the func `releaseResourcesIfPossible()` do actually. Though, not important now.

Reply:
Maybe I missed it, I definitely see the Master doing  releaseResourcesIfPossible which is checking from the Master side whether the worker is alive. But the Master might think its not alive if the network disconnected, but on the Worker it might still be up.  In the normal case Worker will exit because of heartbeat and release its own resources (thus doesn't need this message), if it doesn't exit for some reason then the Worker and its executors could potentially still be alive using the resources.  In this very edge case I would propose we just let it be and don't release the resources. I would hope cluster admin would see this and go figure out what was going on.